### PR TITLE
After Powering on of the node, job goes into Held state

### PR DESCRIPTION
**Pull request by ashwathraop**
#### Bug/feature Description
When node was issued power on and before hook exited we mark node to ND_FREE, which made scheduler to send jobs to those which are still booting up. Jobs try running 21 run times and end up going to held state.

#### Affected Platform(s)
Cray

#### Cause / Analysis / Design
For power on marking nodes to ND_FREE

#### Solution Description
For power on mark nodes to ND_STALE, when node comes up inventroy hook will mark nodes to free.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by bhroam** [comment]
I dislike reusing the stale state.  Up until now it had one meaning.  The mom reported vnodes once and then didn't report it at a later date.  You're actually allowing an admin to mark a vnode as stale.  I dislike this even more.  Stale a PBS owned state.  Why can't you leave the node in the sleep state until it is really up and ready to be used?

**Comment by ashwathraop** [comment]
I have moved node attribute declarations needed by ptl to separate PR https://github.com/PBSPro/pbspro/pull/647

**Comment by bhroam** [comment]
It sounds like this is still the right solution.  Marking the node in the stale state seems like a bad workaround.  How hard would it be to make the inventory hook ramped down aware?

**Comment by ashwathraop** [comment]
@bhroam I reverted the code to allow Stale being set from hook or commands. Now hook will poll for node to come up and removes SLEEP state when node is ready. Also there are some new changes added related to reservations. Please have a look.

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 1]
```diff
@@ -538,4 +538,28 @@ def _pmi_ramp_up(self, hosts):
...
                         sleep_time = random.randint(1,10)
                         time.sleep(sleep_time)
-        return True
\ No newline at end of file
+        return True
+
+    def _pmi_power_status(self, hosts):
+        pbs.logmsg(pbs.LOG_DEBUG, "Cray: status of the nodes")
+        nidset = nodenids(hosts)
+        nids, _ = nidlist(None, nidset)
```
Comment (bhroam) [inline src/modules/python/pbs/v1/_pmi_cray.py]: Instead of passing None as the first argument, can you just pass a named argument for the second?  Or is the first argument required?

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 2]
```diff
@@ -538,4 +538,28 @@ def _pmi_ramp_up(self, hosts):
...
+        else:
+            return nodeset
+        craynid = "PBScraynid"
+        for vnames in hosts:
+            vnode = _svr_vnode(vnames)
+            try:
+                nid = int(vnode.resources_available[craynid])
+                if nid in ready:
+                    nodeset.add(vnames)
+            except Exception:
```
Comment (bhroam) [inline src/modules/python/pbs/v1/_pmi_cray.py]: The only thing that is going to raise an exception is possibly a KeyError.  You don't need the try/except.  You can just say if craynid in vnode.resources_available: above of where you check for it.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 3]
```diff
@@ -287,6 +287,31 @@ if this_event.type == pbs.PERIODIC:
                         exec_vnodes[vn]["neededby"] = start_time
         i += 1
 
+    try:
+        pbs_conf = pbs.get_pbs_conf()
+        pbs_home = pbs_conf['PBS_HOME']
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Don't put this in a try/except.  Just say if 'PBS_HOME' in pbs_conf

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 4]
```diff
@@ -287,6 +287,31 @@ if this_event.type == pbs.PERIODIC:
...
+
+    # Identify the nodes file and read in the data
+    node_file = ''
+    sleep_node_list = []
+    node_file = os.path.join(pbs_home, 'server_priv', 'hooks', 'tmp',
+                              'pbs_power_nodes_file')
+    if os.path.isfile(node_file) and os.stat(node_file).st_size:
+        pbs.logmsg(pbs.EVENT_DEBUG3, "pbs_power_nodes_file is %s" % node_file)
+        try:
+            fd = open(node_file, 'r')
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Use a with statement

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 5]
```diff
@@ -287,6 +287,31 @@ if this_event.type == pbs.PERIODIC:
...
+            sleep_node_list = fd.read().split(',')
+            fd.close()
+        except IOError:
+            raise Exception("I/O error reading node file")
+        except:
+            raise Exception("Error reading node file")
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Don't use a blanket except.  Narrow down the scope.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 6]
```diff
@@ -319,40 +344,73 @@ if this_event.type == pbs.PERIODIC:
...
-                        if resv.state == pbs.RESV_STATE_RUNNING or \
-                                resv.state == pbs.RESV_STATE_DELETED or \
-                                resv.state == pbs.RESV_STATE_BEING_ALTERED or \
-                                resv.state == pbs.RESV_STATE_DELETING_JOBS:
+                for resid in rs_list:
+                    resv = resvlist[resid.lstrip()]
+                    if resv.reserve_state == pbs.RESV_STATE_RUNNING or \
+                            resv.reserve_state == pbs.RESV_STATE_DELETED or \
+                            resv.reserve_state == pbs.RESV_STATE_BEING_ALTERED or \
+                            resv.reserve_state == pbs.RESV_STATE_DELETING_JOBS:
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Instead of this, do
rstates = [pbs.RESV_STATE_RUNNING, pbs.RESV_STATE_DELETED, pbs.RESV_STATE_BEING_ALTERED, pbs.RESV_STATE_DELETING_JOBS]
if resv.reserve_state in rstates:

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 7]
```diff
@@ -319,40 +344,73 @@ if this_event.type == pbs.PERIODIC:
...
+                        nodes[host]["poweroff"] = 0
+                        nodes[host]["can_power_off"] = 0
+                    if resv.reserve_state == pbs.RESV_STATE_CONFIRMED:
+                        reserve_start = resv.reserve_start
+                        if vn not in exec_vnodes:
+                            exec_vnodes[vn] = {}
+                            exec_vnodes[vn]["neededby"] = reserve_start
+                        if reserve_start < exec_vnodes[vn]["neededby"]:
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Can this be an elif?

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 8]
```diff
@@ -319,40 +344,73 @@ if this_event.type == pbs.PERIODIC:
...
+                    continue
+            # Look for upcoming reservation starts in the node.
+            for resid in rs_list:
+                resv = resvlist[resid.lstrip()]
+                if resv.reserve_state == pbs.RESV_STATE_CONFIRMED:
+                    reserve_start = resv.reserve_start
+                    if vn not in exec_vnodes:
+                        exec_vnodes[vn] = {}
+                        exec_vnodes[vn]["neededby"] = reserve_start
+                    if reserve_start < exec_vnodes[vn]["neededby"]:
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Can this be an elif?

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 9]
```diff
@@ -392,23 +450,48 @@ if this_event.type == pbs.PERIODIC:
...
                 power.power_on(poweron_vnlist)
-            up_state = pbs.ND_STALE
+            ready_nodes = set()
+            if sleep_node_list:
+                ready_nodes = power.power_status(sleep_node_list)
+                # For the nodes which are up remove ND_SLEEP
+                for host in ready_nodes:
+                    for vn in nodes[host]["vnodes"]:
+                        prev_state = vn.state
+                        vn.state = prev_state & ~(pbs.ND_SLEEP)
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: I'm a little confused here.  Aren't you just getting the power status of all the nodes in the sleep nodes list?  Couldn't some of them still be coming up?

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 10]
```diff
@@ -392,23 +450,48 @@ if this_event.type == pbs.PERIODIC:
...
+                        vn.last_used_time = time_now
+            # nodes which are still booting, write them to
+            # node file for status check in next iteration.
+            if sleep_node_list or poweron_vnlist:
+                try:
+                    fd = open(node_file, 'w')
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Use a with

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 11]
```diff
@@ -392,23 +450,48 @@ if this_event.type == pbs.PERIODIC:
...
+                    for n in sleep_node_list:
+                        if n not in ready_nodes:
+                            sleeping_nodes.append(n)
+                    data = ','.join(sleeping_nodes)
+                    fd.write(data)
+                    fd.close()
+                except IOError:
+                    raise Exception("I/O error reading node file")
+                except:
+                    raise Exception("Error reading node file")
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Don't use a blanket except.  Narrow down the focus.

File: test/fw/ptl/lib/pbs_api_to_cli.py [hunk 12]
```diff
@@ -278,6 +278,8 @@
     ATTR_NODE_License: 'license',
     ATTR_NODE_LicenseInfo: 'license_info',
     ATTR_NODE_TopologyInfo: 'topology_info',
+    ATTR_NODE_last_used_time: 'last_used_time',
+    ATTR_NODE_last_state_change_time: 'last_state_change_time',
```
Comment (bhroam) [inline test/fw/ptl/lib/pbs_api_to_cli.py]: I thought this was pulled out into a different PR

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 9 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: capmc node_status returns list of nodes based on their state. Something similar to below:
{ready : [1,2,3],
standby: [4,5],
halt: [6,7,8],
on: [9,10],
off: [11,12,13]}
Here nodes which are in ready state only are the usable by PBS. And power.power_status(sleep_node_list) will return only those which are in ready state.

File: test/fw/ptl/lib/pbs_api_to_cli.py (hunk 12 above)
Comment (ashwathraop) [reply test/fw/ptl/lib/pbs_api_to_cli.py]: Rebasing with master messed up my branch somehow. So I did a reset of this file with master. If you look at overall changes you should not see these changes.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 10 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Done.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 11 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Done.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 8 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Okay.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 7 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Okay.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 6 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Okay.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 5 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Done.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 4 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Okay.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 3 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Okay.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 2 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Okay.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 1 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Done. Made first argument too optional and passed a named argument.

File: src/modules/python/pbs/v1/_pmi_cray.py [hunk 13]
```diff
@@ -538,4 +538,26 @@ def _pmi_ramp_up(self, hosts):
                         launch(func, cmd)
                         sleep_time = random.randint(1,10)
                         time.sleep(sleep_time)
-        return True
\ No newline at end of file
+        return True
+
+    def _pmi_power_status(self, hosts):
```
Comment (suresh-thelkar) [inline src/modules/python/pbs/v1/_pmi_cray.py]: Can you add a brief description of the function like the way you are doing it for others.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 14]
```diff
@@ -320,40 +344,73 @@ if this_event.type == pbs.PERIODIC:
...
+                        exec_vnodes[vn] = {}
+                        exec_vnodes[vn]["neededby"] = reserve_start
+                    elif reserve_start < exec_vnodes[vn]["neededby"]:
+                        exec_vnodes[vn]["neededby"] = reserve_start
             if vn in exec_vnodes and exec_vnodes[vn]["neededby"]:
                 neededby = exec_vnodes[vn]["neededby"]
             else:
                 neededby = 0
-            if neededby and neededby < (time_now + freq):
+            if neededby and neededby < (time_now + 900):
```
Comment (suresh-thelkar) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Better to add the significance of 900 here for better readability.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 9 above)
Comment (bhroam) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Oh ok.  The name of that function is a little confusing.  It sounds like it will return the power status of the node list passed to it.

File: test/fw/ptl/lib/pbs_api_to_cli.py (hunk 12 above)
Comment (bhroam) [reply test/fw/ptl/lib/pbs_api_to_cli.py]: Cool thanks.  I took a look and it doesn't show up now.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 15]
```diff
@@ -299,17 +299,16 @@ if this_event.type == pbs.PERIODIC:
...
         pbs.logmsg(pbs.EVENT_DEBUG3, "pbs_power_nodes_file is %s" % node_file)
         try:
-            fd = open(node_file, 'r')
-            sleep_node_list = fd.read().split(',')
-            fd.close()
+            with open(node_file, 'r') as fd:
+                sleep_node_list = fd.read().split(',')
         except IOError:
             raise Exception("I/O error reading node file")
-        except:
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: The python documentation says open() can only raise an IOError.  It raises it when the file can't be opened.  Why are you catching it just to raise a different type of exception?  Just let open raise the exception and don't catch it.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 16]
```diff
@@ -468,19 +468,18 @@ if this_event.type == pbs.PERIODIC:
             # nodes which are still booting, write them to
             # node file for status check in next iteration.
             if sleep_node_list or poweron_vnlist:
+                sleeping_nodes = poweron_vnlist
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Since you are doing an 'or' sleep_node_list or poweron_vlist can be None.  The for or the append() will fail if this is the case.  Are you sure you don't want an 'and'?

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 17]
```diff
@@ -468,19 +468,18 @@ if this_event.type == pbs.PERIODIC:
...
-                            sleeping_nodes.append(n)
-                    data = ','.join(sleeping_nodes)
-                    fd.write(data)
-                    fd.close()
+                    with open(node_file, 'w') as fd:
+                        fd.write(data)
                 except IOError:
                     raise Exception("I/O error reading node file")
-                except:
+                except Exception:
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Same as above, just let the IOError exception be raised uncaught.

File: src/modules/python/pbs/v1/_pmi_cray.py (hunk 13 above)
Comment (ashwathraop) [reply src/modules/python/pbs/v1/_pmi_cray.py]: Okay.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 14 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Okay.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 15 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Sure.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 16 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: We need or as both may not be true all the time. What you said about append() is right so I added an extra check.

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 17 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Done.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 18]
```diff
@@ -308,17 +308,14 @@ if this_event.type == pbs.PERIODIC:
                 sleep_node_list = fd.read().split(',')
         except IOError:
             raise Exception("I/O error reading node file")
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: I'm still unsure why you catch IOError and reraise an exception.  Why not just remove the try/except block and let the IOError go through uncaught.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 19]
```diff
@@ -481,8 +481,6 @@ if this_event.type == pbs.PERIODIC:
                         fd.write(data)
                 except IOError:
                     raise Exception("I/O error reading node file")
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Same here, why don't you just let the IOError go uncaught?

File: src/modules/python/pbs_hooks/PBS_power.PY (hunk 18 above)
Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: I changed the block to reject the hook.

File: src/modules/python/pbs_hooks/PBS_power.PY [hunk 20]
```diff
@@ -306,8 +306,8 @@ if this_event.type == pbs.PERIODIC:
         try:
             with open(node_file, 'r') as fd:
                 sleep_node_list = fd.read().split(',')
-        except IOError:
-            raise Exception("I/O error reading node file")
+        except IOError, e:
```
Comment (bhroam) [inline src/modules/python/pbs_hooks/PBS_power.PY]: Please use the more modern style of exception handling (except IOerror as e).  (same for the rest of the exceptions)

Comment (ashwathraop) [reply src/modules/python/pbs_hooks/PBS_power.PY]: Fixed. Please have a look.

**Comment by suresh-thelkar** [comment]
Code changes look good to me.

**Comment by ashwathraop** [comment]
@bhroam I have squashed and re-based the commits. Please take it forward.

**Comment by ashwathraop** [comment]
@bhroam EDD update is no longer required as we moved away from using Stale state.

**Comment by ashwathraop** [comment]
[ptl.log](https://github.com/PBSPro/pbspro/files/1969411/ptl.log)
PTL power test logs

**Comment by ashwathraop** [comment]
[ptl_smoke.log](https://github.com/PBSPro/pbspro/files/1969436/ptl_smoke.log)
Smoke test logs
